### PR TITLE
Gh-2560 get traits operation configurable

### LINKED ISSUE #2560: Make GetTraits Operation configurable per store
This will mean it no longer relies on the deprecated getTraits function: https://github.com/gchq/Gaffer/blob/296c453ed415d38d7ccd57324089242a2c591a31/core/store/src/main/java/uk/gov/gchq/gaffer/store/Store.java#L359-L360

Instead the operation should be configured per store and setup so that each store's handler relates to the traits of that store.

### PR DESCRIPTION
**Pull request by p3430233**
# Related Issue

- Resolve #2560

### CODE REVIEW

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/StoreTest.java [hunk 1]
```diff
@@ -1150,6 +1152,11 @@ protected void addAdditionalOperationHandlers() {
             return addElementsHandler;
         }
 
+        @Override
+        protected OutputOperationHandler<GetTraits, Set<StoreTrait>> getGetTraitsHandler() {
+            return getTraitsHandler;
```
Comment (t92549) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/StoreTest.java:1157]: This should maybe return an actual handler to retain the same functionality as `getTraits()`

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/StoreTest.java [hunk 2]
```diff
@@ -1245,6 +1252,11 @@ protected void addAdditionalOperationHandlers() {
...
```
Comment (t92549) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/StoreTest.java:1257]: This should maybe return an actual handler to retain the same functionality as `getTraits()`

File: store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedGetTraitsHandlerTest.java [hunk 3]
```diff
@@ -287,6 +287,11 @@ protected void addAdditionalOperationHandlers() {
             return null;
         }
 
+        @Override
+        protected OutputOperationHandler<GetTraits, Set<StoreTrait>> getGetTraitsHandler() {
+            return null;
```
Comment (t92549) [inline store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedGetTraitsHandlerTest.java:292]: This should maybe return an actual handler to retain the same functionality as `getTraits()`

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/operation/handler/TestAddToGraphLibraryImpl.java [hunk 4]
```diff
@@ -61,6 +62,13 @@ protected void addAdditionalOperationHandlers() {
         return null;
     }
 
+    @Override
+    protected OutputOperationHandler<GetTraits, Set<StoreTrait>> getGetTraitsHandler() {
+        Set<StoreTrait> traits = new HashSet<>(StoreTrait.ALL_TRAITS);
+        traits.remove(StoreTrait.ORDERED);
+        return new GetTraitsHandler(traits);
```
Comment (t92549) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/operation/handler/TestAddToGraphLibraryImpl.java:69]: This should return empty and have to be overridden if needed like it did before

**Review by t92549 — CHANGES_REQUESTED** [review]

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java [hunk 5]
```diff
@@ -34,12 +34,17 @@
...
 
+    public GetTraitsHandler(final Set<StoreTrait> storeTraits) {
+        this.storeTraits = storeTraits;
+    }
+
     @Override
     public Set<StoreTrait> doOperation(final GetTraits operation, final Context context, final Store store) throws OperationException {
         if (!operation.isCurrentTraits()) {
-            return store.getTraits();
+            return storeTraits;
```
Comment (GCHQDev404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java:47]: You haven't protected the Set.
+ testing

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java [hunk 6]
```diff
@@ -34,12 +34,17 @@
 import static org.apache.commons.collections.CollectionUtils.isNotEmpty;
 
 public class GetTraitsHandler implements OutputOperationHandler<GetTraits, Set<StoreTrait>> {
+    private final Set<StoreTrait> storeTraits;
     private Set<StoreTrait> currentTraits;
 
+    public GetTraitsHandler(final Set<StoreTrait> storeTraits) {
+        this.storeTraits = storeTraits;
```
Comment (GCHQDev404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java:41]: You haven't protected the set.

**Review by GCHQDev404 — CHANGES_REQUESTED** [review]

**Review by GCHQDeveloper314 — COMMENTED** [review]
The naming of `getGetTraitsHandler` is not the best but it seems to be unavoidable.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java (hunk 6 above)
Comment (t92549) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java]: I believe this has now been fixed

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java (hunk 5 above)
Comment (t92549) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java:47]: I believe this has now been fixed

**Review by t92549 — APPROVED** [review]
I am happy with this but it needs extra approval

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java (hunk 5 above)
Comment (GCHQDev404) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/operation/handler/GetTraitsHandler.java:47]: 👍

**Review by GCHQDev404 — APPROVED** [review]

**Review by GCHQDev404 — APPROVED** [review]

**Pull request merged**